source/core/interface_tools.cc

```diff
@@ -286,6 +334,878 @@ InterfaceTools::SignedDistanceSolver<dim, VectorType>::get_signed_distance()
+
+          double delta_volume_nm1 = 0.0;
+          double delta_volume_n   = 0.0;
+
+          double delta_volume_prime = 0.0;
+
+          double eta_nm1 = 0.0;
+          double eta_n   = 1e-6 * cell_size;
+          double eta_np1 = 0.0;
+
+          // Compute the volume for the first initial value (eta_nm1)
+          inside_cell_volume_nm1 =
+            InterfaceTools::compute_cell_wise_volume(fe_point_evaluation,
+                                                     cell,
+                                                     cell_dof_values,
+                                                     eta_nm1,
+                                                     fe->degree + 1);
+          delta_volume_nm1 = targeted_cell_volume - inside_cell_volume_nm1;
+
+          /* Store the initial volume in the cell to limit the secant method
+          in some case.*/
+          const double initial_inside_cell_volume = inside_cell_volume_nm1;
+
+          /* Check if there is enough volume to correct. If not, we don't
+          correct.*/
+          if (inside_cell_volume_nm1 < 1e-10 * cell_size ||
+              inside_cell_volume_nm1 > (cell_size - 1e-10 * cell_size))
+            {
+              eta_n = 0.0;
+              continue;
+            }
+
+          unsigned int secant_it     = 0;
+          double       secant_update = 1.0;
+          while (abs(secant_update) > 1e-10 &&
+                 abs(delta_volume_nm1) > 1e-10 * initial_inside_cell_volume &&
+                 secant_it < 20)
+            {
+              // If the cell is almost full or empty, we stop correcting the
+              // volume.
+              if (inside_cell_volume_nm1 < 1e-10 * cell_size ||
+                  inside_cell_volume_nm1 > (cell_size - 1e-10 * cell_size))
+                {
+                  eta_n = 0.0;
+                  break;
+                }
+              secant_it += 1;
+
+              inside_cell_volume_n =
+                InterfaceTools::compute_cell_wise_volume(fe_point_evaluation,
+                                                         cell,
+                                                         cell_dof_values,
+                                                         eta_n,
+                                                         fe->degree + 1);
+
+              delta_volume_n = targeted_cell_volume - inside_cell_volume_n;
+
+              delta_volume_prime =
+                (delta_volume_n - delta_volume_nm1) / (eta_n - eta_nm1 + 1e-16);
+
+              secant_update = -delta_volume_n / (delta_volume_prime + 1e-16);
+              eta_np1       = eta_n + secant_update;
+
+              eta_nm1                = eta_n;
+              eta_n                  = eta_np1;
+              inside_cell_volume_nm1 = inside_cell_volume_n;
+              delta_volume_nm1       = delta_volume_n;
+            } // End secant method loop.
+
+          if (secant_it >= 20)
+            {
+              eta_n = 0.0;
+            }
```

Comment:
What does this mean? If you had more than 20 iterations you set it to zero?
Please add a comment here that explains the behaviour.